cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/remote/RemoteExecutionTwillRunnerService.java

```diff
@@ -215,10 +224,22 @@ public TwillPreparer prepare(TwillApplication application) {
     RunId twillRunId = RunIds.generate();
     String remoteHost = masterNode.getProperties().get("ip.external");
 
+    Optional<ProfileId> profileId = SystemArguments.getProfileIdFromArgs(programRunId.getNamespaceId(),
+                                                                         programOptions.getArguments().asMap());
+    if (!profileId.isPresent()) {
+      throw new IllegalStateException(String.format("Missing profile information for this program run %s ",
+                                                    programRunId));
+    }
+    ProfileMetricScheduledService metricScheduledService = new ProfileMetricScheduledService(metricsCollectionService,
```

Comment:
This will start a new thread per run, which is too much. Please refactor the class to take the `monitorScheduler` as parameter and use it as the scheduler (see `AbstractRetryableScheduledService`). Alternatively, you can have your class extends from `AbstractRetryableScheduledService` and override the `executor()` method to simplify the code (see `RuntimeMonitor` as an example).

Reply:
Fixed, now the service will use the `monitorScheduler` as the executor